### PR TITLE
Specialize

### PR DESCRIPTION
**Pull request by atrick**
#### Introduces an @_specialize(...) function attribute.

An internal @_specialize function attribute allows developers to force
full specialization by listing concrete type names corresponding to the
function's generic signature. A function's generic signature is a
concatenation of its generic context and the function's own generic
type parameters.

```
  struct S<T> {
    var x: T
    @_specialize(Int, Float)
    mutating func exchangeSecond<U>(u: U, _ t: T) -> (U, T) {
      x = t
      return (u, x)
    }
  }

  // Substitutes: <T, U> with <Int, Float> producing:
  // S<Int>::exchangeSecond<Float>(u: Float, t: Int) -> (Float, Int)

```

@_specialize currently acts as a hint to the optimizer, which
generates type checks and code to dispatch to the specialized routine
without affecting the signature of the generic function. The
intention is to support efforts at evaluating the performance of
specialized code. The performance impact is not guaranteed and is
likely to change with the optimizer. This attribute should only be
used in conjunction with rigorous performance analysis. Eventually,
a similar attribute could be defined in the language, allowing it to be
exposed as part of a function's API. That would allow direct dispatch
to specialized code without type checks, even across modules.

In the future, we can build on this work in several ways:
- cross module dispatch directly to specialized code
- dynamic dispatch directly to specialized code
- automated specialization based on broader hints
- partial specialization
- and so on...

@swift-ci Please test

### CODE REVIEW

**Comment by atrick** [comment]
@swift-ci Please test

**Comment by swiftix** [comment]
:+1:

File: docs/Generics.rst [hunk 1]
```diff
@@ -713,6 +713,35 @@ the virtual dispatch, inline calls when appropriate, and eliminate the overhead
 of the generic system. Such optimizations can be performed based on heuristics,
 user direction, or profile-guided optimization.
 
+An internal @_specialize function attribute allows devlopers to force
```
Comment (harlanhaskins) [inline docs/Generics.rst]: typo: `devlopers`

File: docs/Generics.rst [hunk 2]
```diff
@@ -713,6 +713,35 @@ the virtual dispatch, inline calls when appropriate, and eliminate the overhead
...
+full specialization by listing concrete type names corresonding to the
```
Comment (harlanhaskins) [inline docs/Generics.rst]: typo: `corresonding`

Comment (atrick) [inline docs/Generics.rst]: Fixed, thanks.

**Comment by swiftix** [comment]
@atrick Are you planning to add a benchmark for this feature? I think it would make a lot of sense.

And, BTW, does this feature generate specializations even at `-Onone`? If it does and it is intended then the tests should check that specializations are produced both for `-Onone` and `-O`

**Comment by atrick** [comment]
@swiftix Yes, I'm planning to add benchmarks. The one that I have is nontrivial to integrate because it involves modules.

I did not tie any of the test cases to the -O or -Onone pipeline. The pass currently runs at -O and I don't have any plans to change -Onone.

**Comment by nadavrot** [comment]
@swift-ci Please test

**Comment by atrick** [comment]
@swift-ci Please test

**Pull request merged**